### ISSUE TITLE
selection-based editors miss lots of assists

### ISSUE THREAD

**Issue by jrvidal** [rust-lang/rust-analyzer#17539]
See https://github.com/helix-editor/helix/issues/9631

I use RA together with Helix. The edition model of this editor is such that the selection is _never empty_. I'm not sure whether this is unique to Helix, maybe Kakoune works in a similar way.

As per the issue linked above, this means that there are many assists that don't appear in their "natural" place, say:

```rust
d[b]g!(some_expr);  // <-- selection of 1 character over "b"
```

does not offer "Remove dbg! macro", because of [this check](https://github.com/rust-lang/rust-analyzer/blob/cae997e3380363a906588f14c7b4587f39cf09f5/crates/ide-assists/src/handlers/remove_dbg.rs#L25). The only possibility is selecting the whole `dbg!(...)` expression.

I'm not sure what the right way to tackle this would be. Maybe there could be a setting that the editor could toggle to indicate "I'm a selection-based editor", so that those assists can relax those checks. Maybe those checks ought to be relaxed in all cases (at least, for selections of length 1)? Maybe this is something the LSP standard should handle itself :see_no_evil: ?

**Comment by jrvidal** [comment]
> we might as well remove the branch, it shouldn't change the behavior

Which branch do you mean in particular?  IIUC, the logic is:

* If no selection, look `node_at_offset()`.
* If there's a selection, check `covering_element()`.

The latter yields nothing for a 1-char selection, so keeping it wouldn't solve the problem. Or did you mean keeping just the first branch?

> I am not sure if any make use of this differentiation right now to change their behavior

FWIW I just tested and `extract_variable` suffers from the same problem.

**Comment by Veykril** [comment]
Right, I misunderstood how the assist worked. This is tricky then, imo there is significant difference between no selection and a single char selection... and some assist make this distinction for a reason.

**Comment by davidbarsky** [comment]
I wonder if it makes sense to extend the assist to span the entire line? It might mean that there'd be more assists, but I've recently found myself slightly annoyed at losing the assist bulb in VS Code because I moved the cursor a character over.

**Comment by matklad** [comment]
Whoa, that's a design bug! Wish I knew about this back then. 

I think the fix is to introduce something like

```rust
fn selected_element(self)
```

to AssistCtx, which internally selects the node based on one of three cases:

1. selection is a pipe cursor `|` (in which case, it needs to correctly handle the case of selection happening exactly _between_ two nodes)
2. selection is a bar cursor `_` (in which case, always, at most one node is selected)
3. selection is a range

That is:

* each individual assist shouldn't care about block vs. pipe
* at the same time, there should be _some_ function that disambiguate the two precisely

I am _not_ sure where there's enough information in LSP to disambiguate between 2. and 3. but in any case this should be trivial to add (it's just a client capability). Even if LSP doesn't tell us this _now_, I think detecting the case heuristically as "selection is one `char`" would work great for 99% of cases. 

---

The second case where this leaks out is snippets, where we treat `$` as betwen characters, but where it might also be _at_ the character .

Fun fact: I am actually using a block cursor in VS Code, and I _was_ annoyed a couple of times that it _incorrectly_ picks an element when it looks like its between character. But I didn't relelize that this actually isn't merely a question of visual representation, but actually affects how assists behave. 

---

I guess the more tactical change is to audit call-sites of `is_empty_selection` and figure out what is the right semantic operation to use. Because, naturally, `is_empty_selection` just doesn't make coherent sense. 

---

would be good to check with upstream LSP spec. Could it be the case that even block-curosr using editors should use "empty" selection to point at a single place in the text buffer?
